src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -298,120 +298,110 @@ internal static bool TryComputeEncodedBytes(ReadOnlySpan<uint> utf32, out int by
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((surrogateCalculation >> 12) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((surrogateCalculation >> 6) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((surrogateCalculation) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
+                                goto need_more; // Invalid surrogate pair.
                             }
-                            else if (codePoint <= Utf8ThreeBytesLastCodePoint)
+
+                            int surrogateCalculation;
+                            unchecked
                             {
-                                if (bytesWrittenTemp > outputBytesAvailable - 3)
-                                {
-                                    goto need_more;
-                                }
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((codePoint >> 12) & b0000_1111U) | b1110_0000U);
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((codePoint >> 6) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((codePoint) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
+                                surrogateCalculation = ((lowSurrogate - Utf16LowSurrogateFirstCodePoint)
+                                    | ((codePoint - Utf16HighSurrogateFirstCodePoint) << 10)) + 0x10000;
                             }
-                            else
+
+                            *(bytesWrittenTemp + pUtf8) = (byte)(((surrogateCalculation >> 18) & b0000_0111U) | b1111_0000U);
+                            bytesWrittenTemp++;
+                            *(bytesWrittenTemp + pUtf8) = (byte)(((surrogateCalculation >> 12) & b0011_1111U) | b1000_0000U);
+                            bytesWrittenTemp++;
+                            *(bytesWrittenTemp + pUtf8) = (byte)(((surrogateCalculation >> 6) & b0011_1111U) | b1000_0000U);
+                            bytesWrittenTemp++;
+                            *(bytesWrittenTemp + pUtf8) = (byte)(((surrogateCalculation) & b0011_1111U) | b1000_0000U);
+                            bytesWrittenTemp++;
+                        }
+                        else
+                        {
+                            if (bytesWrittenTemp > outputBytesAvailable - 3)
                             {
-                                throw new ArgumentException("Invalid surrogate pair.", nameof(utf16));
+                                goto need_more;
```

Comment:
As per the suggestion, I added a test case where the input buffer only contains low surrogates to ensure correctness in this scenario. There is also a test that contains a high surrogate which isn't paired with a low surrogate.